File: memory-daemon/claudia_memory/services/audit.py
"""
Audit Service for Claudia Memory System

Tracks all operations for debugging, accountability, and compliance.
"""

import json
import logging
from datetime import datetime
from typing import Any, Dict, List, Optional

from ..database import get_db

logger = logging.getLogger(__name__)


class AuditService:
    """Track all memory system operations"""

    def __init__(self):
        self.db = get_db()

    def log(
        self,
        operation: str,
        details: Optional[Dict[str, Any]] = None,
        session_id: Optional[str] = None,
        user_initiated: bool = False,
        entity_id: Optional[int] = None,
        memory_id: Optional[int] = None,
    ) -> int:
        """
        Log an operation to the audit trail.

        Args:
            operation: Operation type (e.g., 'entity_merge', 'memory_correct', 'entity_delete')
            details: JSON-serializable details about the operation
            session_id: Optional session identifier
            user_initiated: Whether this was triggered by user action
            entity_id: Optional entity ID this operation affects
            memory_id: Optional memory ID this operation affects

        Returns:
            Audit log entry ID
        """
        entry_id = self.db.insert(
            "audit_log",
            {
                "timestamp": datetime.utcnow().isoformat(),
                "operation": operation,
                "details": json.dumps(details) if details else None,
                "session_id": session_id,
                "user_initiated": 1 if user_initiated else 0,
                "entity_id": entity_id,
                "memory_id": memory_id,
            },
        )
        logger.debug(f"Audit logged: {operation} (id={entry_id})")
        return entry_id

    def get_recent(
        self,
        limit: int = 50,
        operation: Optional[str] = None,
        entity_id: Optional[int] = None,
        memory_id: Optional[int] = None,
    ) -> List[Dict[str, Any]]:
        """
        Get recent audit entries.

        Args:
            limit: Maximum entries to return
            operation: Filter by operation type
            entity_id: Filter by entity ID
            memory_id: Filter by memory ID

        Returns:
            List of audit entries (newest first)
        """
        sql = "SELECT * FROM audit_log WHERE 1=1"
        params = []

        if operation:
            sql += " AND operation = ?"
            params.append(operation)

        if entity_id:
            sql += " AND entity_id = ?"
            params.append(entity_id)

        if memory_id:
            sql += " AND memory_id = ?"
            params.append(memory_id)

        sql += " ORDER BY timestamp DESC LIMIT ?"
        params.append(limit)

        rows = self.db.execute(sql, tuple(params), fetch=True) or []
        return [
            {
                "id": row["id"],
                "timestamp": row["timestamp"],
                "operation": row["operation"],
                "details": json.loads(row["details"]) if row["details"] else None,
                "session_id": row["session_id"],
                "user_initiated": bool(row["user_initiated"]),
                "entity_id": row["entity_id"],
                "memory_id": row["memory_id"],
            }
            for row in rows
        ]

    def get_entity_history(self, entity_id: int) -> List[Dict[str, Any]]:
        """
        Get all audit entries affecting an entity.

        Args:
            entity_id: Entity ID to trace

        Returns:
            List of audit entries (oldest first)
        """
        rows = self.db.execute(
            """
            SELECT * FROM audit_log
            WHERE entity_id = ?
            ORDER BY timestamp ASC
            """,
            (entity_id,),
            fetch=True,
        ) or []
        return [
            {
                "id": row["id"],
                "timestamp": row["timestamp"],
                "operation": row["operation"],
                "details": json.loads(row["details"]) if row["details"] else None,
                "session_id": row["session_id"],
                "user_initiated": bool(row["user_initiated"]),
            }
            for row in rows
        ]

    def get_memory_history(self, memory_id: int) -> List[Dict[str, Any]]:
        """
        Get all audit entries affecting a memory.

        Args:
            memory_id: Memory ID to trace

        Returns:
            List of audit entries (oldest first)
        """
        rows = self.db.execute(
            """
            SELECT * FROM audit_log
            WHERE memory_id = ?
            ORDER BY timestamp ASC
            """,
            (memory_id,),
            fetch=True,
        ) or []
        return [
            {
                "id": row["id"],
                "timestamp": row["timestamp"],
                "operation": row["operation"],
                "details": json.loads(row["details"]) if row["details"] else None,
                "session_id": row["session_id"],
                "user_initiated": bool(row["user_initiated"]),
            }
            for row in rows
        ]


# Global service instance
_service: Optional[AuditService] = None


def get_audit_service() -> AuditService:
    """Get or create the global audit service"""
    global _service
    if _service is None:
        _service = AuditService()
    return _service


# Convenience functions
def audit_log(operation: str, **kwargs) -> int:
    """Log an operation to the audit trail"""
    return get_audit_service().log(operation, **kwargs)


def get_audit_recent(limit: int = 50, **kwargs) -> List[Dict[str, Any]]:
    """Get recent audit entries"""
    return get_audit_service().get_recent(limit, **kwargs)


def get_entity_audit_history(entity_id: int) -> List[Dict[str, Any]]:
    """Get all audit entries affecting an entity"""
    return get_audit_service().get_entity_history(entity_id)


def get_memory_audit_history(memory_id: int) -> List[Dict[str, Any]]:
    """Get all audit entries affecting a memory"""
    return get_audit_service().get_memory_history(memory_id)

File: cli/services/audit.js
/**
 * Audit service for Claudia CLI.
 * Port of memory-daemon/claudia_memory/services/audit.py.
 *
 * Tracks all memory system operations for debugging and accountability.
 */

/**
 * Log an operation to the audit trail.
 * @param {object} db - ClaudiaDatabase instance
 * @param {string} operation - Operation type (e.g., 'entity_merge', 'memory_correct')
 * @param {object} [options]
 * @param {object} [options.details] - JSON-serializable details
 * @param {string} [options.sessionId] - Session identifier
 * @param {boolean} [options.userInitiated] - Whether triggered by user
 * @param {number} [options.entityId] - Entity ID this affects
 * @param {number} [options.memoryId] - Memory ID this affects
 * @returns {number} Audit log entry ID
 */
export function auditLog(db, operation, { details, sessionId, userInitiated, entityId, memoryId } = {}) {
  try {
    return db.insert('audit_log', {
      timestamp: new Date().toISOString(),
      operation,
      details: details ? JSON.stringify(details) : null,
      session_id: sessionId || null,
      user_initiated: userInitiated ? 1 : 0,
      entity_id: entityId || null,
      memory_id: memoryId || null,
    });
  } catch {
    // Audit logging should never crash the caller
    return -1;
  }
}

/**
 * Get recent audit entries.
 * @param {object} db
 * @param {object} [options]
 * @param {number} [options.limit=50]
 * @param {string} [options.operation]
 * @param {number} [options.entityId]
 * @param {number} [options.memoryId]
 * @returns {object[]}
 */
export function getAuditRecent(db, { limit = 50, operation, entityId, memoryId } = {}) {
  let sql = 'SELECT * FROM audit_log WHERE 1=1';
  const params = [];

  if (operation) {
    sql += ' AND operation = ?';
    params.push(operation);
  }
  if (entityId) {
    sql += ' AND entity_id = ?';
    params.push(entityId);
  }
  if (memoryId) {
    sql += ' AND memory_id = ?';
    params.push(memoryId);
  }

  sql += ' ORDER BY timestamp DESC LIMIT ?';
  params.push(limit);

  return (db.query(sql, params) || []).map(formatAuditEntry);
}

/**
 * Get all audit entries affecting an entity.
 * @param {object} db
 * @param {number} entityId
 * @returns {object[]}
 */
export function getEntityAuditHistory(db, entityId) {
  return (db.query(
    'SELECT * FROM audit_log WHERE entity_id = ? ORDER BY timestamp ASC',
    [entityId],
  ) || []).map(formatAuditEntry);
}

/**
 * Get all audit entries affecting a memory.
 * @param {object} db
 * @param {number} memoryId
 * @returns {object[]}
 */
export function getMemoryAuditHistory(db, memoryId) {
  return (db.query(
    'SELECT * FROM audit_log WHERE memory_id = ? ORDER BY timestamp ASC',
    [memoryId],
  ) || []).map(formatAuditEntry);
}

function formatAuditEntry(row) {
  return {
    id: row.id,
    timestamp: row.timestamp,
    operation: row.operation,
    details: row.details ? JSON.parse(row.details) : null,
    session_id: row.session_id,
    user_initiated: Boolean(row.user_initiated),
    entity_id: row.entity_id,
    memory_id: row.memory_id,
  };
}
